code_generation/code_generation.py

```diff
@@ -1,29 +1,42 @@
-import code_generation.producer as p
+class SafeDict(dict):
+    def __missing__(self, key):
+        return "{" + key + "}"
 
 
 def fill_template(t, config):
-    # shift producers
-    for shift in config.keys():
-        # shift names start with "_"
-        if not shift.startswith("_"):
-            continue
-        for entry in config[shift]["shiftbase"]:
-            getattr(p, entry).shift(shift)
-
     # generate list of commands
     commandlist = ""  # string to be placed into code template
     df_count = 0  # enumerate dataframes
+    df_scope_count = 0  # enumerate dataframes in certain scopes
 
     # get commands of producers and append to the command list
-    for producer in config["producers"]:
-        commandlist += "\n    //" + producer + "\n"
-        for call in getattr(p, producer).writecalls(config):
-            # print(call)
+    for producer in config["producers"]["global"]:
```

Comment:
Couldnt this function be put in the for loop below, there we iterate on global anyawys

Reply:
It must be done before all other scopes and I think the dict sorts alphabetically and would change the order given in the config.